include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -1752,656 +1572,197 @@ namespace internal
-    AssertIndexRange(direction, dim);
-    constexpr int mm = contract_over_rows ? n_rows : n_columns,
-                  nn = contract_over_rows ? n_columns : n_rows;
-
-    constexpr int stride    = Utilities::pow(n_columns, direction);
-    constexpr int n_blocks1 = one_line ? 1 : stride;
-
-    // The number of blocks depend on both direction and dimension.
-    constexpr int n_blocks2 =
-      (dim - direction - 1 == 0) ?
-        1 :
-        ((direction == normal_dir) ?
-           Utilities::pow((n_rows - 1),
-                          (direction >= dim) ? 0 : dim - direction - 1) :
-           (((direction < normal_dir) ? (n_rows + 1) : n_rows) *
-            ((dim - direction == 3) ? n_rows : 1)));
-
-    for (int i2 = 0; i2 < n_blocks2; ++i2)
-      {
-        for (int i1 = 0; i1 < n_blocks1; ++i1)
-          {
-            Number x[mm];
-            for (int i = 0; i < mm; ++i)
-              x[i] = in[stride * i];
-
-            for (int col = 0; col < nn; ++col)
-              {
-                Number2 val0;
-
-                if (contract_over_rows)
-                  val0 = shape_data[col];
-                else
-                  val0 = shape_data[col * n_columns];
-
-                Number res0 = val0 * x[0];
-                for (int i = 1; i < mm; ++i)
-                  {
-                    if (contract_over_rows)
-                      val0 = shape_data[i * n_columns + col];
-                    else
-                      val0 = shape_data[col * n_columns + i];
-
-                    res0 += val0 * x[i];
-                  }
-                if (add)
-                  out[stride * col] += res0;
-
-                else
-                  out[stride * col] = res0;
-              }
-
-            if (one_line == false)
-              {
-                ++in;
-                ++out;
-              }
-          }
-        if (one_line == false)
-          {
-            in += stride * (mm - 1);
-            out += stride * (nn - 1);
-          }
-      }
-  }
-
-  template <int dim,
-            int n_rows,
-            int n_columns,
-            int normal_dir,
-            typename Number,
-            typename Number2>
-  template <int  face_direction,
+  template <int  n_rows_template,
+            int  stride_template,
             bool contract_onto_face,
             bool add,
-            int  max_derivative>
-  inline void
-  EvaluatorTensorProductAnisotropic<
-    evaluate_raviart_thomas,
-    dim,
-    n_rows,
-    n_columns,
-    normal_dir,
-    Number,
-    Number2>::apply_face(const Number *DEAL_II_RESTRICT in,
-                         Number *DEAL_II_RESTRICT       out) const
+            int  max_derivative,
+            typename Number,
+            typename Number2>
+  inline std::enable_if_t<!contract_onto_face, void>
+  interpolate_to_face(const Number2            *shape_values,
+                      const std::array<int, 2> &n_blocks,
+                      const std::array<int, 2> &steps,
```

Comment:
I think this makes sense to be more general. But for a user to use this function becomes probably more complicated. Maybe some comments here would help: what are blocks and steps?

Reply:
I have documentation in the first of the two functions, explaining what it does. I agree it gets more complicated at the user side, but it really allows me to use the same interpolation in essentially all cases, be it isotropic or anisotropic. What the other function says with my update:
```
   * @param n_blocks Number of interpolation layers used along the up to two
   *             dimensions tangential to the interpolation direction.
   * @param steps Increments in the input array from one step to the next,
   *             varied in conjunction with the @p stride_template variable: We
   *             increment by @p stride_template along the 1d interpolation,
   *             and then increment by @p steps when passing from one line
   *             to the next.
```
So from this you can see that we only do this up to 3D because we assume at most two tangential directions (did we use this in hyper.deal directly?), albeit one could argue that all dimensions can be implemented in tensor contractions by two loops wrapping the dimensions before and after the 1d interpolation of sum factorization, to be compared against a single outer loop in matrix-matrix multiplications.

Reply:
> albeit one could argue that all dimensions can be implemented in tensor contractions by two loops wrapping the dimensions before and after the 1d interpolation of sum factorization, to be compared against a single outer loop in matrix-matrix multiplications.

I know. But to see this one needs to write the nested loops once and then collapse them :+1: 

> (did we use this in hyper.deal directly?),

I don't think so. But if we did, I will update the code the code there. It is an internal function.